Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/safetynet/SafetyNetClientWrapper.kt

```diff
@@ -0,0 +1,116 @@
+package de.rki.coronawarnapp.datadonation.safetynet
+
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.common.api.CommonStatusCodes
+import com.google.android.gms.safetynet.SafetyNetApi
+import com.google.android.gms.safetynet.SafetyNetClient
+import com.google.gson.JsonObject
+import com.google.gson.JsonParser
+import dagger.Reusable
+import de.rki.coronawarnapp.datadonation.safetynet.SafetyNetException.Type
+import de.rki.coronawarnapp.environment.EnvironmentSetup
+import okio.ByteString.Companion.decodeBase64
+import timber.log.Timber
+import javax.inject.Inject
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+@Reusable
+class SafetyNetClientWrapper @Inject constructor(
+    private val safetyNetClient: SafetyNetClient,
+    private val environmentSetup: EnvironmentSetup
+) {
+
+    suspend fun attest(nonce: ByteArray): Report {
+        val jwsResult = callClient(nonce).jwsResult ?: throw SafetyNetException(
+            Type.ATTESTATION_FAILED,
+            "JWS was null"
+        )
+
+        val components = jwsResult.split(".")
+        if (components.size != 3) throw SafetyNetException(
+            Type.ATTESTATION_FAILED,
+            "Invalid JWS: Components are missing."
+        )
+
+        val header = try {
+            components[0].decodeBase64Json()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS header.", e)
+        }
+        val body = try {
+            components[1].decodeBase64Json()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS body.", e)
+        }
+
+        val signature = try {
+            components[2].decodeBase64()!!.toByteArray()
+        } catch (e: Exception) {
+            throw SafetyNetException(Type.ATTESTATION_FAILED, "Failed to decode JWS signature.", e)
+        }
+
+        return Report(
+            jwsResult = jwsResult,
+            header = header,
+            body = body,
+            signature = signature
+        )
+    }
+
+    private fun String.decodeBase64Json(): JsonObject {
+        val rawJson = decodeBase64()!!.string(Charsets.UTF_8)
+        return JsonParser.parseString(rawJson).asJsonObject
+    }
+
+    private suspend fun callClient(nonce: ByteArray): SafetyNetApi.AttestationResponse = suspendCoroutine { cont ->
+        safetyNetClient.attest(nonce, environmentSetup.safetyNetApiKey)
+            .addOnSuccessListener {
+                Timber.tag(TAG).v("Attestation finished with %s", it)
+                cont.resume(it)
+            }
+            .addOnFailureListener {
+                Timber.tag(TAG).w(it, "Attestation failed.")
```

Comment:
This [checklist](https://developer.android.com/training/safetynet/attestation#use-response-server) contains: 

> Your app handles transient errors by retrying the request with an increasing amount of time between retries (exponential backoff).

Should we implement a retry mechanism?

Reply:
I'm not sure we can make the decision at this level. For analytics we can retry a lot and also delay for longer, it's in the background. For the survey stuff the user is active on screen. Both cases likely need different retry behavior.

I'd say this is up to the classes consuming the API?

Reply:
True, there are different circumstances in which our API is called...
Then it makes sense to leave it to the consumers of the API.